Assets/MixedRealityToolkit.Services/InputSystem/InputSystemGlobalHandlerListener.cs

```diff
@@ -0,0 +1,164 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using Microsoft.MixedReality.Toolkit.Utilities;
+using System;
+using System.Reflection;
+using System.Threading.Tasks;
+using UnityEngine;
+using UnityEngine.EventSystems;
+
+namespace Microsoft.MixedReality.Toolkit.Input
+{
+    /// <summary>
+    /// This component ensures that input events are forwarded to this component when focus or gaze is not required.
+    /// </summary>
+    public abstract class InputSystemGlobalHandlerListener : MonoBehaviour
+    {
+        private bool lateInitialize = true;
+
+        private readonly Type eventSystemHandlerType = typeof(IEventSystemHandler);
+        private readonly MethodInfo registerHandlerMethod = typeof(IMixedRealityEventSystem).GetMethod("RegisterHandler");
+        private readonly MethodInfo unregisterHandlerMethod = typeof(IMixedRealityEventSystem).GetMethod("UnregisterHandler");
+
+        private IMixedRealityInputSystem inputSystem = null;
+
+        /// <summary>
+        /// The active instance of the input system.
+        /// </summary>
+        protected IMixedRealityInputSystem InputSystem
+        {
+            get
+            {
+                if (inputSystem == null)
+                {
+                    MixedRealityServiceRegistry.TryGetService<IMixedRealityInputSystem>(out inputSystem);
+                }
+                return inputSystem;
+            }
+        }
+
+        protected virtual void OnEnable()
+        {
+            if (InputSystem != null && !lateInitialize)
+            {
+                RegisterHandlers();
+            }
+        }
+
+        protected virtual async void Start()
```

Comment:
>void [](start = 32, length = 4)

As long as we're adding a new file, it might be worth considering the definition here (or the structure via which sub-classes end up consuming this).

In particular, because this is async void, subclasses actually won't be able to await base.Start() because Start is void.

Then again, it's not really possible to mark this as async Task, because then it wouldn't match Unity's Start() definition (I don't think it will get called at that point - I'm not 100% on this part).

The only REAL thing woudl be to have like this other virtual AwaitingStart() function that looks like:

protected virtual async void Start()
{
   ... start stuff...
   await AwaitingStart();
}

protected virtual async Task AwaitingStart()
{
}

The other possibility is to actually make this sealed or not extendable (but leave a comment saying it would be possible by doing what is mentioned above)

I guess the main point here is that there was a deficiency in the old thing which, as long as we're adding new thing, we may not want to choose to carry over.

Reply:
Do we actually have a scenario where we'd want a delayed registration in `Start` and not just fail fast in `enable`, if input system is not present? 

---
In reply to: [295375494](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/4970#discussion_r295375494) [](ancestors = 295375494)

Reply:
>Do we actually have a scenario where we'd want a delayed registration in Start and not just fail fast in enable, if input system is not present?

By that argument, should the base class Start need to be async at all? :)

Reply:
My question was more into "do we actually need to wait until `Start`, do we have a case when input system is not available in `OnEnable`? Do we want such cases to be supported at all if they happen?". I haven't seen all parts of the system yet, but it looks like we don't. CC @davidkline-ms @wiwei

Reply:
I don't have enough background to say authoritatively.

@davidkline-ms, it has to be async because there is an await there.

I think my thought here is, if we just make this private, we can sidestep this question until someone needs it, at which point we can do the right thing.

Reply:
Right thing being, do some virtual task-returning thingy